<div class="panel panel-primary">
    <div class="panel-heading" >
        <strong>miRNASNP-v3 document</strong>
	</div>
	<div class="panel-body">
		<ul>
		<li><a href="#!/document/#one">1. What's new in miRNASNP-v3 compared to the previous version?</a></li>
		<li><a href="#!/document/#two">2. Data statistics for miRNASNP-v3</a></li>
		<!--li><a href="#!/document/#three">3. Pipeline to construct miRNASNP3</a></li-->
		<li><a href="#!/document/#three">3. Data sources and tools for analyze</a></li>
		<li><a href="#!/document/#four">4. Phenotype</a></li>
		<li><a href="#!/document/#five">5. Enrichment analysis for gained/lost targets</a></li>
		<li><a href="#!/document/#six">6. miRNA information and gene 3'UTR</a></li>
		<li><a href="#!/document/#seven">7. Expression level and expression correlation of miRNAs and target genes</a></li>
		<li><a href="#!/document/#eight">8. Genome-wide association studies (GWAS)</a></li>
		<li><a href="#!/document/#nine">9. Online prediction</a></li>
		<li><a href="#!/document/#ten">10. Target gain/loss</a></li>
		<li><a href="#!/document/#eleven">11. Cite miRNASNP3</a></li>
		<li><a href="#!/document/#contact">12. Contact us</a></li>
		</ul>
		<div id="one">
			<p>
				<span class="document">
					What's new in miRNASNP-v3 compared to the previous version?
				</span>
			</p>
			<!--p>
					microRNAs (miRNAs) are endogenous and regulatory non-coding RNAs by targeting mRNAs for cleavage or translational repression. 
					SNPs on pre-miRNAs or target sites will affect miRNA expression or functions. 
					miRNASNP is a comprehensive resource of human miRNA-related SNPs, including their effects on target gain and loss.
			</p-->
			<p>
					As the number of SNPs included in the dbSNP has increased significantly, and many new miRNAs have been discovered in the past few years, 
					we released miRNASNP-v3. In this new version, several new features are added, including:
			</p>
			<ul>
				<li><span class="glyphicon glyphicon-star" aria-hidden="true" style="font: 2px"></span> phenotypes for variations located in pre-miRNA and gene 3'UTR,</li>
				<li><span class="glyphicon glyphicon-star" aria-hidden="true"></span> analysis of target gain/loss caused by variations in miRNA seed and gene 3'UTR,</li>
				<li><span class="glyphicon glyphicon-star" aria-hidden="true"></span> gene enrichment analysis of miRNA targets gain and loss.</li>
			</ul>
			<!--p>And We update data to the latest version, User can</p>
			<ul>
				<li>search or browse SNPs in pre-miRNA and their impacts on miRNAs' structure,</li>
				<li>search or browse targets gain/loss by SNP in miRNA seed region and gene 3'UTR,</li>
				<li>search or browse experiment validated miRNA:gene interaction,</li>
				<li>search or browse expression and correlation of miRNAs and genes,</li>
				<li>search or browse expression and correlation of miRNAs and genes.</li>
			</ul-->
		</div>
		<div id="two">
				<p>
					<span class="document">
						Data statistics for miRNASNP-v3
					</span>
				</p>
				<!--p>
						In total, There are 17,361 SNPs and 2,155 variations identified in mature miRNA. 43,923 SNPs and 5,464 variations are identified in pre-miRNA.
						The number of SNPs and variations in gene 3'UTR is 6,358,867 and 202,942 separately.
				</p-->
				<div align="center">
				<img src="/static/miRNASNP3/img/data_statistics_table.png" style="margin-top: 30px;margin-bottom: 30px;width: 80%">
				</div>
				<!--div align="center">
						<img src="/static/miRNASNP3/img/targets_statistics_1.png" style="margin-top: 30px;margin-bottom: 30px">
					</div-->
			</div>
		<!--div id="three">
			<p>
				<span class="document">
						Pipeline to construct miRNASNP3
				</span>
			</p>
			<div align="center">
							<img src="/static/miRNASNP3/img/pipline-1.png" style="margin-top: 30px;width: 800px;">
							</div>
		</div-->
		<div id="three">
			<p>
				<span class="document">Data sources and tools for analyze</span>
			</p>
			<table class="table table-bordered table-striped table-hover">
				<tbody>
					<tr>
						<th style="width:25%">Data</th>
						<th style="width:25%">Source</th>
						<th style="width:50%">Url</th>
					</tr>
					<tr>
						<td>miRNA</td>
						<td>miRBase v22</td>
						<td><a href="http://www.mirbase.org">http://www.mirbase.org</a></td>
					</tr>
					<tr>
						<td>SNP</td>
						<td>dbSNP v151</td>
						<td><a href="https://www.ncbi.nlm.nih.gov/snp/">https://www.ncbi.nlm.nih.gov/snp/</a></td>
					</tr>
					<tr>
						<td>GWAS SNPs</td>
						<td>GWAS catelog</td>
						<td><a href="https://www.ebi.ac.uk/gwas/home">https://www.ebi.ac.uk/gwas/home</a></td>
					</tr>
					<tr>
						<td rowspan="2">Gene</td>
						<td>UCSC</td>
						<td><a href="https://genome.ucsc.edu/cgi-bin/hgTables">https://genome.ucsc.edu/cgi-bin/hgTables</a></td>
					</tr>
					<tr>
						<td>Ensembl</td>
						<td><a href="http://www.ensembl.org/biomart/martview">http://www.ensembl.org/biomart/martview</a></td>
					</tr>
					<tr>
						<td>miRNA and gene expression</td>
						<td>TCGA</td>
						<td><a href="https://portal.gdc.cancer.gov/">https://portal.gdc.cancer.gov/</a></td>
					</tr>
					<tr>
					<td>COSMIC Noncoding Variation</td>
					<td>COSMIC</td>
					<td><a href="https://cancer.sanger.ac.uk/cosmic/">https://cancer.sanger.ac.uk/cosmic/</a></td>
					</tr>
					<tr>
						<td>ClinVar variation</td>
						<td>ClinVar</td>
						<td><a href="https://www.ncbi.nlm.nih.gov/clinvar/">https://www.ncbi.nlm.nih.gov/clinvar/</a></td>
					</tr>
					<tr>
						<td rowspan="5">Experiment validated</td>
						<td>TarBase v8</td>
						<td><a href="http://carolina.imis.athena-innovation.gr/diana_tools/web/index.php?r=tarbasev8">http://carolina.imis.athena-innovation.gr/diana_tools/web/index.php?r=tarbasev8</a></td>
					</tr>
					<tr>
						<td>miR2Disease</td>
						<td><a href="http://www.mir2disease.org/">http://www.mir2disease.org/</a></td>
					</tr>
					<tr>
						<td>miRecords v4</td>
						<td><a href="http://c1.accurascience.com/miRecords/">http://c1.accurascience.com/miRecords/</a></td>
					</tr>
					<tr>
						<td>miRTarBase v7</td>
						<td><a href="http://mirtarbase.mbc.nctu.edu.tw/">http://mirtarbase.mbc.nctu.edu.tw/</a></td>
					</tr>
					<tr>
						<td>StarBase v3</td>
						<td><a href="http://www.sysu.edu.cn/">http://www.sysu.edu.cn/</a></td>
					</tr>
					
				</tbody>
			</table>
			<table class="table table-bordered table-striped table-hover">
					<tbody>
						<tr>
							<th style="width:25%">Tools</th>
							<th style="width:25%">Usage</th>
							<th style="width:50%">Url</th>
						</tr>
						<tr>
							<td>TargetScan</td>
							<td>miRNA target prediction</td>
							<td><a href="http://www.targetscan.org/vert_72/">http://www.targetscan.org/vert_72/</a></td>
						</tr>
						<tr>
							<td>miRmap</td>
							<td>miRNA target prediction</td>
							<td><a href="https://mirmap.ezlab.org/">https://mirmap.ezlab.org/</a></td>
						</tr>
						<tr>
							<td>VennaRNA</td>
							<td>pre-miRNA second structure prediction </td>
							<td><a href="https://www.tbi.univie.ac.at/RNA/index.html">https://www.tbi.univie.ac.at/RNA/index.html</a></td>
						</tr>
						</tbody>
						</table>
		</div>

		
		<div id="four">
			<p>
				<span class="document">
							Phenotype
						</span>
						</p>
			<p>
					In the phenotype-related variations module, we integrated pathological information of 175,985 mutations from <a href="https://cancer.sanger.ac.uk/cosmic">COSMIC</a>, disease information of 32,136 variations from <a href="https://www.ncbi.nlm.nih.gov/clinvar/">ClinVar</a>
					 and disease/trait information of 235,148 SNPs from <a href="https://www.ebi.ac.uk/gwas/">GWAS catalog</a>.
					We divided these variations into three categories based on their location. For variation in miRNA seed regions and gene 3'UTRs, target gain/loss information was presented.
					For variations in pre-miRNAs, we provided the minimum free energy change of pre-miRNA second structure caused by the variation, users can click the pre-miRNA to look into pre-miRNA details.
					In this module, users can screen for variations associated with a particular phenotype by selecting a tissue of interest or typing in a description of the phenotype.
			</p>
			<div align="center">
				<img src="/static/miRNASNP3/img/phenotype-help.png" style="width: 100%;margin-top: 30px;margin-bottom: 30px">
			</div>
				
		</div>
		
		<div id="five">
				<p>
						<span class="document">
									Enrichment analysis for gained/lost targets
								</span>
								</p>
				<p>
						Enrichment analysis of gained/lost targets by variations in miRNA seed region may help us further discover the underlying biological mechanisms effected by the variation.
						We perform <a href="http://geneontology.org/">Gene Ontology (GO)</a> and <a href="https://www.kegg.jp/">Kyoto Encyclopedia of Genes and Genomes (KEGG)</a> enrichment analysis on gained/lost gene sets with gene numbers from 50 to 1000.
				</p>
					<div align="center">
						<img src="/static/miRNASNP3/img/enrichment-help.png" style="width: 100%;margin-top: 30px;margin-bottom: 30px">
					</div>
		</div>

		<div id="six">
			<p>
			<span class="document">
				miRNA information and gene 3'UTR
			</span>
			</p>
			<p>
					The host gene information of miRNAs was extracted from mirna_context.txt file.
					For miRNAs without host gene, we label them as intergenic.
					The gene 3'UTR file and human genome sequence(GRCh38)  were downloaded by <a href="http://genome.ucsc.edu/cgi-bin/hgTables">the UCSC Table browser</a>, then we use bedtools to get the sequence 
					information for each transcript based on the genome position. To get a more reliable transcript set, 
					gene symbols and <a href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> mRNA ID corresponding to the transcript ID was downloaded from <a href="http://asia.ensembl.org/biomart/martview/">Ensembl</a>, transcripts which not have gene symbol or <a href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> mRNA ID
					were excluded. Finally, 34,116 transcripts(transcripts with a sequence length greater than 6(seed region length)) of 18,286 genes were ultimately used for target analysis.
			</p>
		</div>

		<div id="seven">
				<p>
				<span class="document">
					Expression level and expression correlation of miRNAs and target genes
				</span>
				</p>
			<p>
					Gene expression data and miRNA expression data across 33 cancers were downloaded from <a href="https://portal.gdc.cancer.gov/">TCGA</a>.
					For each cancer, We performed pearson correlation analysis between gene and miRNA.
					The overall correlation coefficient value was displayed in loss targets information panel, which can be clicked to pop out a histogram of correlation coefficient value for each cancer type. 
					However, because there is no expression data for SNP type 3'UTR/miRNA, the correlation coefficient value for gain targets is not available.
			</p>
			<div align="center">
				<img style="width: 100%;margin-top: 30px;margin-bottom: 30px" src="/static/miRNASNP3/img/correlation_and_expression.png">
			</div>
			
			</div>
		
	<div id="eight">
		<p>
			<span class="document">Genome-wide association studies (GWAS)</span>
		</p>
		<p>
				GWAS tagSNPs and SNPs in  linkage disequilibrium(LD) region of the tagSNP were downloaded from <a href="https://www.ebi.ac.uk/gwas/">GWAS catalog</a> website for 26 populations. 
				Then, an integrated LD region for tagSNPs related to a specific SNP of each population is calculated by SNPs in tagSNP LD regions.
				(setting LD analysis regions &plusmn;250kb around SNP position)
		</p>
		<div align="center">
				<img style="width: 100%;margin-top: 30px" src="/static/miRNASNP3/img/ld_3_top.png">
				<img style="width: 100%;margin-bottom: 30px" src="/static/miRNASNP3/img/ld_3_buttom.png">
				<!--img style="width: 100%;margin-bottom: 30px;margin-top: 30px" src="/static/miRNASNP3/img/ld_2.png"-->
			</div>
	</div>

	<div id="nine">
		<p>
			<span class="document">Online prediction</span>
		</p>
		<p>
				miRNASNP-v3 provided 3 kinds of online predictions. Predicting the impact of variant in gene 3'UTR on miRNA:mRNA interaction will scan 2,652 mature miRNAs
    to predict target miRNAs for wild-type and SNP-type 3'UTR sequence, which provided by users, by <a href="https://mirmap.ezlab.org/">miRmap</a> and <a href="http://www.targetscan.org/vert_72/">TargetScan</a>. Then follow the analysis pipeline constructed in
    miRNASNP to get the gain/loss targets. Predicting the impact of variant in miRNA on miRNA:mRNA interaction will scan 18,286 gene 3'UTRs to predict target genes
    for wild-type and SNP-type miRNA sequence.(In order to reduce the calculation time, only the longest transcript of each gene was used to take online prediction in Tools module.)
    The prediction result will provide a summary panel, and another four panels to show gain targets, loss targets, wild-type and SNP-type sequence predicted targets. For predicting the impact of variations in pre-miRNA on miRNA structure, the
    minimum free energy(MFE) of SNP-type sequence for each SNP is predicted by<a href="https://www.tbi.univie.ac.at/RNA/index.html">RNAfold</a>.
		</p>
		<div align="center">
				<img src="/static/miRNASNP3/img/predict-help.png" style="width: 100%;margin-top: 30px;margin-bottom: 30px">
			</div>
	</div>

	<div id="ten">
			<p>
				<span class="document">Target gain/loss</span>
			</p>
			<p>We provide the effect of SNPs/variations in miRNA seed regions. 
				For SNPs in miRNA seed region, we got miRNA wild sequence and SNP allele sequence. 
				Then, we used two target prediction tools <a href="http://www.targetscan.org/vert_72/">TargetScan</a> and <a href="https://mirmap.ezlab.org/">miRmap</a> to predict their target sites respectively. 
				Four categories of results are recorded as wild <a href="http://www.targetscan.org/vert_72/">TargetScan</a> site (WT), wild <a href="https://mirmap.ezlab.org/">miRmap</a> site (WM), SNP <a href="http://www.targetscan.org/vert_72/">TargetScan</a> site (ST) and SNP <a href="https://mirmap.ezlab.org/">miRmap</a> site (SM). If one miRNA's target gene shows in both WT and WM, but not in either ST or SM, we called that the miRNA lost this target gene. 
				On the contrary, if one target shows in both ST and SM, but not in either WT or WM, we called the miRNA gained one target gene.</p>
				<div align="center">
						<img src="/static/miRNASNP3/img/target_mirna.png" style="width: 80%;margin-top: 30px;margin-bottom: 30px">
					</div>
			<p>We also provides the effect of SNPs in genes' 3'UTRs. 
				For SNPs in one gene UTR, we got wild type 3'UTR sequence and corresponding SNP 3'UTR sequence. 
				Then, we used two target prediction tools (<a href="http://www.targetscan.org/vert_72/">TargetScan</a> and <a href="https://mirmap.ezlab.org/">miRmap</a>) to predict their target sites respectively. 
				Four categories of results are recorded as wild <a href="http://www.targetscan.org/vert_72/">TargetScan</a> site (WT), wild <a href="https://mirmap.ezlab.org/">miRmap</a> site (WM), SNP <a href="http://www.targetscan.org/vert_72/">TargetScan</a> site (ST), SNP <a href="https://mirmap.ezlab.org/">miRmap</a> site (SM). If one target site shows in both WT and WM, but not in either ST or SM, we called that the gene lost this target site. 
				On the contrary, if one target both show in ST and SM, but not in either WT or WM, we called the gene gained one target site.</p>
				<div align="center">
						<img src="/static/miRNASNP3/img/target_geneutr.png" style="width: 80%;margin-top: 30px;margin-bottom: 30px">
					</div>
	</div>

	<div id="eleven">
		<p>
			<span class="document">Cite miRNASNP</span>
		</p>
		<p>
				<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4397995/" target="_blank">
					<strong>An update of
						miRNASNP database for better SNP selection by GWAS data, miRNA expression and online
				tools.
					</strong>
				</a> <br />
						Gong J, Liu CJ, Wu Y, Ma Z, Chen H and Guo AY. <em>Database</em>. (2015)
			</p>
			<p>
					<a href="https://onlinelibrary.wiley.com/doi/full/10.1002/humu.21641" target="_blank">
						<strong>
							Genome-wide identification of SNPs in microRNA genes and the SNP effects
							on microRNA target binding and biogenesis.
						</strong>
					</a><br />
							Gong J, Tong Y, Zhang HM, Wang K, Hu T, Shan G, Sun J and Guo AY. <em>Human Mutation</em>. (2011)
				</p>
	</div>

	<div id="contact">
		<p>
			<span class="document">Contact us</span>
		</p>
		<address>
				<strong>&nbsp;&nbsp;An-Yuan Guo, Ph.D. Professor of Bioinformatics</strong><br />
				<strong style="font-size: 18px;margin-top: 1px" >&nbsp;&nbsp;Email:&nbsp</strong>&nbsp;
			   <img alt="An-Yuan Guo" src="/static/miRNASNP3/img/guoay_email.png" style="height: 20px">
			</address>
			<address>
				<strong>&nbsp;&nbsp;Xin Fu, M.S. student</strong><br />
				<strong style="font-size: 18px;margin-top: 1px" >&nbsp;&nbsp;Email:&nbsp</strong>&nbsp;
			   <img alt="Gui-Yan Xie" src="/static/miRNASNP3/img/fuxin_email.png" style="height: 20px">
			</address>
			<address>
				<strong>&nbsp;&nbsp;Chun-Jie Liu, Ph.D. Postdoc Fellow</strong><br/>
				<strong style="font-size: 18px;margin-top: 1px">&nbsp;&nbsp;Email:&nbsp</strong>&nbsp;
				<img alt="Chun-Jie Liu"src="/static/miRNASNP3/img/chunjie_email.png" style="height: 20px">
			</address>
	
		  <div class="address">
				&nbsp;&nbsp;<a href="http://bioinfo.life.hust.edu.cn/" target=_blank">Hubei Bioinformatics & Molecular Imaging Key Laboratory</a> <br />
				&nbsp;&nbsp;Department of Bioinformatics and Systems Biology<br />
				&nbsp;&nbsp;<a href="http://life.hust.edu.cn/" target=_blank">College of Life Science and Technology</a><br />
				&nbsp;&nbsp;<a href="http://www.hust.edu.cn/" target=_blank">Huazhong University of Science and Technology, Wuhan 430074 P.R. China</a><br />
			</div>
	</div>
		
	</div>
</div>		

<!--div class="panel-body" data-spy="scroll" data-target="#documentspy">
		<div class="row">
			<div class="col-lg-2" id="documentspy"> 
							
			<div id="navbutton" class="btn-group-vertical" style="position:fixed;text-align:left">
				<button type="button" class="btn btn-default" id="intro" ng-click="gotoAnchor(1)" style="color: darkred;">Introduction</button>
				<button type="button" class="btn btn-default" id="qs" ng-click="gotoAnchor(2)" style="color: darkred;">Quick Start</button>
				<button type="button" class="btn btn-default" id="ds" ng-click="gotoAnchor(3)" style="color: darkred;">Data Summary</button>
			</div>
				
			</div>
			
			
			<div class="col-lg-10">
				<div id="anchor1">
					<span class="glyphicon glyphicon-paperclip" style="display:inline-block"></span>&nbsp;<h2 style="display:inline-block">Introduction</h2>
					<p>Introduce miRNASNP3</p>
					<b>Abstract summary</b>
					<img src="static/miRNASNP3/img/miRNASNP_frame.png" width="90%" class="img-responsive" style="width: 300px;height:300px">
				</div>
				
				
				<div id="anchor2">
					<span class="glyphicon glyphicon-import" style="display:inline-block"></span>&nbsp;<h2 style="display:inline-block">Quick Start</h2>
					<div id="usage">
						<div id="browseusage">
							<h3>Browse</h3>
							<ol>
								<li><p>Browse miRNASNP3</p></li>
								
							</ol>
						</div>
						<div id="searchusage">
							<h3>Search</h3>
							<ol>
								<li><p>Search</p></li>
							</ol>
						</div>
					</div>
				</div>
				
				<div id="anchor3">
					<span style="display:inline-block" class="glyphicon glyphicon-paperclip"></span>&nbsp;<h2 style="display:inline-block">Data Summary</h2>
					<img src="static/miRNASNP3/img/data_static.png" width="90%" class="img-responsive" style="width: 300px;height:300px">	
				</div>
							
			
			<div id="anchor4">
				<span style="display:inline-block" class="glyphicon glyphicon-paperclip"></span>&nbsp;<h2 style="display:inline-block">Contact</h2>
							<address>
								<strong>&nbsp;&nbsp;An-Yuan Guo, Ph.D. Professor of Bioinformatics</strong><br />
								<strong style="font-size: 18px;margin-top: 1px" >&nbsp;&nbsp;Email:&nbsp</strong>&nbsp;
							   <img alt="An-Yuan Guo" src="/static/miRNASNP3/img/guoay_email.png" style="height: 20px">
							</address>
							<address>
								<strong>&nbsp;&nbsp;Xin Fu, M.S. student</strong><br />
								<strong style="font-size: 18px;margin-top: 1px" >&nbsp;&nbsp;Email:&nbsp</strong>&nbsp;
							   <img alt="Gui-Yan Xie" src="/static/miRNASNP3/img/fuxin_email.png" style="height: 20px">
							</address>
							<address>
								<strong>&nbsp;&nbsp;Chun-Jie Liu, Ph.D. Postdoc Fellow</strong><br/>
								<strong style="font-size: 18px;margin-top: 1px">&nbsp;&nbsp;Email:&nbsp</strong>&nbsp;
								<img alt="Chun-Jie Liu"src="/static/miRNASNP3/img/chunjie_email.png" style="height: 20px">
							</address>
					
						  <div class="address">
								&nbsp;&nbsp;<a href="http://bioinfo.life.hust.edu.cn/" target=_blank">Hubei Bioinformatics & Molecular Imaging Key Laboratory</a> <br />
								&nbsp;&nbsp;Department of Bioinformatics and Systems Biology<br />
								&nbsp;&nbsp;<a href="http://life.hust.edu.cn/" target=_blank">College of Life Science and Technology</a><br />
								&nbsp;&nbsp;<a href="http://www.hust.edu.cn/" target=_blank">Huazhong University of Science and Technology, Wuhan 430074 P.R. China</a><br />
							</div>
				</div>
			
		
				
					<div id="anchor5">
						<span class="glyphicon glyphicon-paperclip" style="display:inline-block"></span>&nbsp;<h2 style="display:inline-block">Citation</h2>
							<div class="col-md-11">
								<p>
									<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4397995/" target="_blank">
										<strong>An update of
											miRNASNP database for better SNP selection by GWAS data, miRNA expression and online
									tools.
										</strong>
									</a> <br />
											Gong J, Liu CJ, Wu Y, Ma Z, Chen H and Guo AY. <em>Database</em>. (2015)
								</p>
							</div>
							<div class="col-md-11 ">
								<p>
									<a href="https://onlinelibrary.wiley.com/doi/full/10.1002/humu.21641" target="_blank">
										<strong>
											Genome-wide identification of SNPs in microRNA genes and the SNP effects
											on microRNA target binding and biogenesis.
										</strong>
									</a><br />
											Gong J, Tong Y, Zhang HM, Wang K, Hu T, Shan G, Sun J and Guo AY. <em>Human Mutation</em>. (2011)
								</p>
					</div>
				</div>
		</div>
	</div-->
		
					
	

